File: scripts/parsePdfWithVision.ts
/**
 * PDF Rules Parser - Pure TypeScript with PDF.js + GPT-5 Vision
 * 
 * Uses Mozilla's PDF.js to render PDFs (no system dependencies!)
 * Then uses GPT-5-mini Vision to extract stratagem text.
 * 
 * Features:
 * - Pure TypeScript/Node.js - no Ghostscript needed
 * - High-quality rendering with PDF.js
 * - GPT-5-mini Vision for accurate text extraction
 * - Generic extraction (preserves all text)
 * 
 * Usage:
 *   npx tsx scripts/parsePdfWithVision.ts --input data/pdf-source/space-wolves.pdf --faction "Space Marines" --detachment "Saga of the Beastslayer"
 */

import * as fs from 'fs-extra';
import * as path from 'path';
import OpenAI from 'openai';
import cliProgress from 'cli-progress';
import { createCanvas } from 'canvas';
import * as pdfjsLib from 'pdfjs-dist/legacy/build/pdf.mjs';

const openai = new OpenAI({ apiKey: process.env.OPENAI_API_KEY });

const MARKDOWN_OUTPUT_DIR = path.join(process.cwd(), 'data', 'rules-source');

// ============================================
// PDF to Image Conversion (Pure JS)
// ============================================

async function convertPdfPageToImage(
  pdfPath: string,
  pageNumber: number
): Promise<string> {
  // Load PDF document
  const data = new Uint8Array(await fs.readFile(pdfPath));
  const loadingTask = pdfjsLib.getDocument({ data });
  const pdfDoc = await loadingTask.promise;
  
  // Get specific page
  const page = await pdfDoc.getPage(pageNumber);
  
  // Set up canvas with high resolution
  const scale = 2.0; // 2x for high quality
  const viewport = page.getViewport({ scale });
  
  const canvas = createCanvas(viewport.width, viewport.height);
  const context = canvas.getContext('2d');
  
  // Render page to canvas
  const renderContext = {
    canvasContext: context as any,
    viewport: viewport
  };
  
  await page.render(renderContext).promise;
  
  // Convert canvas to base64 PNG
  const base64Image = canvas.toDataURL('image/png').split(',')[1];
  
  return base64Image;
}

async function getTotalPages(pdfPath: string): Promise<number> {
  const data = new Uint8Array(await fs.readFile(pdfPath));
  const loadingTask = pdfjsLib.getDocument({ data });
  const pdfDoc = await loadingTask.promise;
  return pdfDoc.numPages;
}

// ============================================
// Vision OCR with GPT-5
// ============================================

async function extractTextFromImage(
  base64Image: string,
  pageNumber: number,
  context: { faction?: string; detachment?: string; }
): Promise<string> {
  const imageUrl = `data:image/png;base64,${base64Image}`;
  
  const systemPrompt = `You are an expert at extracting text from Warhammer 40,000 rulebook pages.

Your job is to extract ALL text content from this page accurately, preserving the exact wording and structure.

EXTRACTION RULES:
1. Extract ALL visible text from the page
2. Preserve exact wording - do not paraphrase, summarize, or interpret
3. Maintain the original structure and formatting where possible
4. If you see stratagems, abilities, or rules, extract them completely
5. Include rule names, costs (CP), descriptions, and any metadata
6. If the page contains no game rules (e.g., table of contents, index, artwork only), respond with "NO_RULES_FOUND"
7. Use clear section breaks between different rules or entries

OUTPUT FORMAT:
- Use markdown headings (##) for rule/stratagem names
- Keep CP costs in parentheses after the name
- Preserve all rule text exactly as written
- Use line breaks to separate different sections
- Include any section headers, rule categories, or metadata you see

${context.faction ? `Context: This is from ${context.faction} content` : ''}
${context.detachment ? `Context: This is from the ${context.detachment} detachment` : ''}

Focus on accuracy and completeness. The extracted text will be parsed and structured later.`;

  try {
    const response = await openai.responses.create({
      model: 'gpt-5-mini',
      instructions: systemPrompt,
      input: [
        {
          role: 'user',
          content: [
            {
              type: 'input_text',
              text: 'Extract all text from this rulebook page. Preserve exact wording and structure.'
            },
            {
              type: 'input_image',
              image_url: imageUrl,
              detail: 'high'
            }
          ]
        }
      ],
    });
    
    const extractedText = response.output_text || '';
    
    if (extractedText.includes('NO_RULES_FOUND')) {
      console.log(`  ⏭️  Page ${pageNumber}: No rules (skipped)`);
      return '';
    }
    
    console.log(`  ✓ Page ${pageNumber}: Text extracted (${extractedText.length} chars)`);
    return extractedText;
    
  } catch (error) {
    console.error(`❌ Error processing page ${pageNumber}:`, error);
    return '';
  }
}

// ============================================
// Main Processing Function
// ============================================

async function processPdfToMarkdown(
  pdfPath: string,
  outputPath: string,
  context: { faction?: string; detachment?: string; }
): Promise<void> {
  console.log('\n' + '='.repeat(60));
  console.log('📚 PDF Parser - Pure TypeScript with GPT-5 Vision');
  console.log('='.repeat(60));
  console.log(`Input:  ${pdfPath}`);
  console.log(`Output: ${outputPath}`);
  if (context.faction) console.log(`Faction: ${context.faction}`);
  if (context.detachment) console.log(`Detachment: ${context.detachment}`);
  console.log('='.repeat(60));
  
  // Get total pages
  const totalPages = await getTotalPages(pdfPath);
  console.log(`\n📄 PDF has ${totalPages} page(s)`);
  
  // Process each page
  const allExtractedText: string[] = [];
  const progressBar = new cliProgress.SingleBar({
    format: 'Processing Pages |{bar}| {percentage}% | {value}/{total} pages',
    barCompleteChar: '\u2588',
    barIncompleteChar: '\u2591',
  });
  
  progressBar.start(totalPages, 0);
  
  for (let pageNum = 1; pageNum <= totalPages; pageNum++) {
    try {
      // Convert page to image
      const base64Image = await convertPdfPageToImage(pdfPath, pageNum);
      
      // Extract text with Vision API
      const extracted = await extractTextFromImage(base64Image, pageNum, context);
      
      if (extracted) {
        allExtractedText.push(extracted);
      }
      
      progressBar.update(pageNum);
      
      // Rate limiting - wait between API calls
      if (pageNum < totalPages) {
        await new Promise(resolve => setTimeout(resolve, 1000)); // 1 second delay
      }
    } catch (error) {
      console.error(`\n❌ Error on page ${pageNum}:`, error);
      progressBar.update(pageNum);
    }
  }
  
  progressBar.stop();
  
  // Combine and write to markdown file
  const header = `# ${context.detachment || context.faction || 'Game'} Rules

Extracted from: ${path.basename(pdfPath)}
Date: ${new Date().toLocaleDateString()}
Total Pages: ${totalPages}
Pages with Content: ${allExtractedText.length}
${context.faction ? `Faction: ${context.faction}` : ''}
${context.detachment ? `Detachment: ${context.detachment}` : ''}

Note: This is raw extracted text. Run through importStrategicRules.ts to parse and structure.

---

`;
  
  const markdownContent = header + allExtractedText.join('\n\n---\n\n');
  
  await fs.ensureDir(path.dirname(outputPath));
  await fs.writeFile(outputPath, markdownContent, 'utf-8');
  
  // Summary
  console.log('\n' + '='.repeat(60));
  console.log('✅ PDF Extraction Complete!');
  console.log('='.repeat(60));
  console.log(`📊 Statistics:`);
  console.log(`   - Total pages: ${totalPages}`);
  console.log(`   - Pages with content: ${allExtractedText.length}`);
  console.log(`   - Total characters: ${markdownContent.length.toLocaleString()}`);
  console.log(`   - Output file: ${outputPath}`);
  console.log('');
  console.log('Next steps:');
  console.log(`   1. Review extracted text: ${outputPath}`);
  console.log(`   2. Edit/clean up if needed`);
  console.log(`   3. Import with AI parser:`);
  console.log(`      npx tsx scripts/importStrategicRules.ts --file ${path.basename(outputPath)}${context.faction ? ` --faction "${context.faction}"` : ''}${context.detachment ? ` --detachment "${context.detachment}"` : ''}`);
  console.log('');
}

// ============================================
// CLI Entry Point
// ============================================

async function main() {
  const args = process.argv.slice(2);
  
  // Parse arguments
  const getArg = (flag: string) => {
    const index = args.indexOf(flag);
    return index !== -1 ? args[index + 1] : null;
  };
  
  const inputPdf = getArg('--input');
  const outputMd = getArg('--output');
  const faction = getArg('--faction');
  const detachment = getArg('--detachment');
  
  // Validate arguments
  if (!inputPdf) {
    console.error('❌ Missing required argument: --input');
    console.log('\nUsage:');
    console.log('  npx tsx scripts/parsePdfWithVision.ts --input <pdf-file> [--output <markdown-file>] [--faction <name>] [--detachment <name>]');
    console.log('\nExamples:');
    console.log('  npx tsx scripts/parsePdfWithVision.ts --input data/pdf-source/space-wolves.pdf --faction "Space Marines" --detachment "Saga of the Beastslayer"');
    console.log('  npx tsx scripts/parsePdfWithVision.ts --input data/pdf-source/tyranids.pdf --faction "Tyranids" --detachment "Subterranean Assault"');
    console.log('  npx tsx scripts/parsePdfWithVision.ts --input data/pdf-source/core-rules.pdf --output core-stratagems.md');
    process.exit(1);
  }
  
  // Resolve paths
  const pdfPath = path.isAbsolute(inputPdf) ? inputPdf : path.join(process.cwd(), inputPdf);
  
  if (!await fs.pathExists(pdfPath)) {
    console.error(`❌ PDF file not found: ${pdfPath}`);
    process.exit(1);
  }
  
  // Auto-generate output filename if not specified
  const defaultOutput = path.join(
    MARKDOWN_OUTPUT_DIR,
    path.basename(inputPdf, '.pdf') + '.md'
  );
  const outputPath = outputMd 
    ? (path.isAbsolute(outputMd) ? outputMd : path.join(process.cwd(), outputMd))
    : defaultOutput;
  
  // Process PDF
  await processPdfToMarkdown(pdfPath, outputPath, {
    faction,
    detachment
  });
}

// ============================================
// Run
// ============================================

main()
  .catch((error) => {
    console.error('❌ Fatal error:', error);
    process.exit(1);
  });

File: scripts/parsePdfWithVision.py
#!/usr/bin/env python3
"""
PDF Rules Parser - Python with PyMuPDF + GPT-5 Vision

Pure Python solution for extracting rules from Warhammer 40K PDFs.
Uses PyMuPDF (no external dependencies) + OpenAI Vision API.

Installation:
    pip install pymupdf openai pillow

Usage:
    python scripts/parsePdfWithVision.py --input data/pdf-source/space-wolves.pdf --faction "Space Marines" --detachment "Saga of the Beastslayer"
"""

import argparse
import base64
import json
import os
import sys
from pathlib import Path
from typing import Optional
import fitz  # PyMuPDF
from openai import OpenAI
from tqdm import tqdm

client = OpenAI(api_key=os.environ.get("OPENAI_API_KEY"))

def convert_pdf_page_to_image(pdf_path: str, page_num: int) -> str:
    """
    Convert a PDF page to base64-encoded PNG image.
    
    Args:
        pdf_path: Path to PDF file
        page_num: Page number (0-indexed)
    
    Returns:
        Base64-encoded PNG image string
    """
    doc = fitz.open(pdf_path)
    page = doc.load_page(page_num)
    
    # Render at 2x resolution (300 DPI) for better OCR
    mat = fitz.Matrix(2.0, 2.0)
    pix = page.get_pixmap(matrix=mat)
    
    # Convert to PNG bytes
    img_bytes = pix.tobytes("png")
    
    # Encode to base64
    base64_image = base64.b64encode(img_bytes).decode('utf-8')
    
    doc.close()
    return base64_image

def extract_text_from_image(
    base64_image: str,
    page_num: int,
    faction: Optional[str] = None,
    detachment: Optional[str] = None
) -> str:
    """
    Extract text from image using GPT-5-mini Vision API.
    """
    system_prompt = f"""You are an expert at extracting text from Warhammer 40,000 rulebook pages.

Your job is to extract ALL text content from this page accurately, preserving the exact wording and structure.

EXTRACTION RULES:
1. Extract ALL visible text from the page
2. Preserve exact wording - do not paraphrase, summarize, or interpret
3. Maintain the original structure and formatting where possible
4. If you see stratagems, abilities, or rules, extract them completely
5. Include rule names, costs (CP), descriptions, and any metadata
6. If the page contains no game rules (e.g., table of contents, index, artwork only), respond with "NO_RULES_FOUND"
7. Use clear section breaks between different rules or entries

OUTPUT FORMAT:
- Use markdown headings (##) for rule/stratagem names
- Keep CP costs in parentheses after the name
- Preserve all rule text exactly as written
- Use line breaks to separate different sections
- Include any section headers, rule categories, or metadata you see

{f"Context: This is from {faction} content" if faction else ""}
{f"Context: This is from the {detachment} detachment" if detachment else ""}

Focus on accuracy and completeness. The extracted text will be parsed and structured later."""

    try:
        response = client.responses.create(
            model="gpt-5-mini",
            instructions=system_prompt,
            input=[
                {
                    "role": "user",
                    "content": [
                        {
                            "type": "input_text",
                            "text": "Extract all text from this rulebook page. Preserve exact wording and structure."
                        },
                        {
                            "type": "input_image",
                            "image_url": f"data:image/png;base64,{base64_image}",
                            "detail": "high"
                        }
                    ]
                }
            ],
        )
        
        extracted_text = response.output_text or ""
        
        if "NO_RULES_FOUND" in extracted_text:
            print(f"  Page {page_num + 1}: No rules (skipped)")
            return ""
        
        print(f"  Page {page_num + 1}: Text extracted ({len(extracted_text)} chars)")
        return extracted_text
        
    except Exception as e:
        print(f"ERROR processing page {page_num + 1}: {e}")
        return ""

def process_pdf(
    pdf_path: str,
    output_path: str,
    faction: Optional[str] = None,
    detachment: Optional[str] = None
):
    """
    Process PDF and extract all text using Vision API.
    """
    print("\n" + "=" * 60)
    print("PDF Parser - Python with GPT-5 Vision")
    print("=" * 60)
    print(f"Input:  {pdf_path}")
    print(f"Output: {output_path}")
    if faction:
        print(f"Faction: {faction}")
    if detachment:
        print(f"Detachment: {detachment}")
    print("=" * 60)
    
    # Get total pages
    doc = fitz.open(pdf_path)
    total_pages = len(doc)
    doc.close()
    
    print(f"\nPDF has {total_pages} page(s)")
    
    # Process each page with progress bar
    all_extracted = []
    
    for page_num in tqdm(range(total_pages), desc="Processing pages", unit="page"):
        try:
            # Convert page to image
            base64_image = convert_pdf_page_to_image(pdf_path, page_num)
            
            # Extract text with Vision API
            extracted = extract_text_from_image(base64_image, page_num, faction, detachment)
            
            if extracted:
                all_extracted.append(extracted)
            
            # Rate limiting
            import time
            if page_num < total_pages - 1:
                time.sleep(1)  # 1 second delay between pages
                
        except Exception as e:
            print(f"\nError on page {page_num + 1}: {e}")
            continue
    
    # Create markdown output
    header = f"""# {detachment or faction or 'Game'} Rules

Extracted from: {Path(pdf_path).name}
Date: {Path(output_path).stat().st_mtime if Path(output_path).exists() else 'N/A'}
Total Pages: {total_pages}
Pages with Content: {len(all_extracted)}
{"Faction: " + faction if faction else ""}
{"Detachment: " + detachment if detachment else ""}

Note: This is raw extracted text. Run through importStrategicRules.ts to parse and structure.

---

"""
    
    markdown_content = header + "\n\n---\n\n".join(all_extracted)
    
    # Write to file
    Path(output_path).parent.mkdir(parents=True, exist_ok=True)
    Path(output_path).write_text(markdown_content, encoding='utf-8')
    
    # Summary
    print("\n" + "=" * 60)
    print("PDF Extraction Complete!")
    print("=" * 60)
    print(f"Statistics:")
    print(f"   - Total pages: {total_pages}")
    print(f"   - Pages with content: {len(all_extracted)}")
    print(f"   - Total characters: {len(markdown_content):,}")
    print(f"   - Output file: {output_path}")
    print("")
    print("Next steps:")
    print(f"   1. Review extracted text: {output_path}")
    print(f"   2. Edit/clean up if needed")
    print(f"   3. Import with AI parser:")
    faction_arg = f' --faction "{faction}"' if faction else ''
    detachment_arg = f' --detachment "{detachment}"' if detachment else ''
    print(f"      npx tsx scripts/importStrategicRules.ts --file {Path(output_path).name}{faction_arg}{detachment_arg}")
    print("")

def main():
    parser = argparse.ArgumentParser(
        description="Extract Warhammer 40K rules from PDF using GPT-5 Vision"
    )
    parser.add_argument('--input', required=True, help='Input PDF file path')
    parser.add_argument('--output', help='Output markdown file path (auto-generated if not specified)')
    parser.add_argument('--faction', help='Faction name (e.g., "Space Marines", "Tyranids")')
    parser.add_argument('--detachment', help='Detachment name')
    
    args = parser.parse_args()
    
    # Resolve paths
    pdf_path = Path(args.input)
    if not pdf_path.is_absolute():
        pdf_path = Path.cwd() / pdf_path
    
    if not pdf_path.exists():
        print(f"ERROR: PDF file not found: {pdf_path}")
        sys.exit(1)
    
    # Auto-generate output filename
    if args.output:
        output_path = Path(args.output)
        if not output_path.is_absolute():
            output_path = Path.cwd() / output_path
    else:
        output_dir = Path.cwd() / 'data' / 'rules-source'
        output_path = output_dir / (pdf_path.stem + '.md')
    
    # Process PDF
    process_pdf(
        str(pdf_path),
        str(output_path),
        args.faction,
        args.detachment
    )

if __name__ == '__main__':
    main()

